## Exploratory Data Analysis

## Phase 1 – Data Understanding & Quality Assessment

In [45]:
import pandas as pd


df = pd.read_csv(
    r"C:\Code\Python\DA_Projects\ecommerce_customer_sales_analytics\dataset\data.csv",
    encoding="latin1")

df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


In [46]:
print(df.dtypes) # Datentypen der Spalten herausfinden

InvoiceNo          str
StockCode          str
Description        str
Quantity         int64
InvoiceDate        str
UnitPrice      float64
CustomerID     float64
Country            str
dtype: object


InvoiceNo: transaction number (multiple products can have the same)

StockCode: product no


In [47]:
# change date from string to numpy date
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"], format="%m/%d/%Y %H:%M")

print(df["InvoiceDate"].head())
print(type(df["InvoiceDate"].iloc[0]))

0   2010-12-01 08:26:00
1   2010-12-01 08:26:00
2   2010-12-01 08:26:00
3   2010-12-01 08:26:00
4   2010-12-01 08:26:00
Name: InvoiceDate, dtype: datetime64[us]
<class 'pandas.Timestamp'>


### Q1. How many missing values are there in each column?

In [48]:
print(df.isnull().sum())

print(df.isnull().mean() * 100)

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64
InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64


We have 135080 missing CustomerID's, that's around 25% of all orders have no CustomerID and 0.26% have no description.

### Q2. How many completely duplicated rows are in the dataset?

In [49]:
duplicates = df.duplicated().sum()
print(f"Duplicated rows: {duplicates}")

Duplicated rows: 5268


There are 5,268 completely duplicated rows in the dataset.

### Q3. What time period does the dataset cover?

In [50]:
earliest_date = df["InvoiceDate"].min()

latest_date = df["InvoiceDate"].max()

print(f"Data covers: {earliest_date} to {latest_date}")

Data covers: 2010-12-01 08:26:00 to 2011-12-09 12:50:00


### Q4. How many transactions have negative Quantity?

In [51]:
negative_quantity_df = df[df["Quantity"] < 0].copy()

print(f"There are {len(negative_quantity_df)} rows with negative quantity")

negative_quantity_df


There are 10624 rows with negative quantity


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom


### Q5. How many transactions have a UnitPrice of 0?

In [52]:
df_unitprice_zero = df[df["UnitPrice"] == 0].copy()

print(f"There are {len(df_unitprice_zero)} transactions with a UnitPrice of 0")

There are 2515 transactions with a UnitPrice of 0


### Q6: How many transactions are cancelled invoices?

In [53]:
negative_quantity_df.head(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom


--> wee can see, that canceled transactions have a "C" at the beginning of the InvoiceNo

In [54]:
canceled_trans_no = df["InvoiceNo"].str.startswith("C").sum()

print(f"There are {canceled_trans_no} canceled transactions")

There are 9288 canceled transactions


In [55]:
negative_not_canceled = df[
    (df["Quantity"] < 0) &
    (~df["InvoiceNo"].str.startswith("C"))
    ]

print(f"There are {len(negative_not_canceled)} negative Quantity transactions but not canceled")

There are 1336 negative Quantity transactions but not canceled


We found that there are 10,624 transactions with a negative quantity. Of these, 9,288 are cancelled invoices, while 1,336 have a negative quantity but are not marked as cancelled.

### Q7. What percentage of all transactions are cancelled?

In [56]:
canceled_percent = (canceled_trans_no / len(df))

print(f"{canceled_percent:.2%} are cancelled transactions")

1.71% are cancelled transactions


Cancelled invoices account for 1.71% of all transactions in the dataset.

### Q8. Are there products with unusually high quantities or prices?

In [57]:
df.sort_values(by="Quantity", ascending=False).head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
540421,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom
61619,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,United Kingdom
502122,578841,84826,ASSTD DESIGN 3D PAPER STICKERS,12540,2011-11-25 15:57:00,0.00,13256.0,United Kingdom
74614,542504,37413,NaN,5568,2011-01-28 12:03:00,0.00,NaN,United Kingdom
421632,573008,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,4800,2011-10-27 12:26:00,0.21,12901.0,United Kingdom
206121,554868,22197,SMALL POPCORN HOLDER,4300,2011-05-27 10:52:00,0.72,13135.0,United Kingdom
220843,556231,85123A,?,4000,2011-06-09 15:04:00,0.00,NaN,United Kingdom
97432,544612,22053,EMPIRE DESIGN ROSETTE,3906,2011-02-22 10:43:00,0.82,18087.0,United Kingdom
270885,560599,18007,ESSENTIAL BALM 3.5g TIN IN ENVELOPE,3186,2011-07-19 17:04:00,0.06,14609.0,United Kingdom
160546,550461,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-04-18 13:20:00,2.10,15749.0,United Kingdom


In [58]:
df.sort_values(by="UnitPrice", ascending=False).head(60)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
222681,C556445,M,Manual,-1,2011-06-10 15:31:00,38970.00,15098.0,United Kingdom
524602,C580605,AMAZONFEE,AMAZON FEE,-1,2011-12-05 11:36:00,17836.46,NaN,United Kingdom
43702,C540117,AMAZONFEE,AMAZON FEE,-1,2011-01-05 09:55:00,16888.02,NaN,United Kingdom
43703,C540118,AMAZONFEE,AMAZON FEE,-1,2011-01-05 09:57:00,16453.71,NaN,United Kingdom
16356,C537651,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:49:00,13541.33,NaN,United Kingdom
15016,C537630,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:04:00,13541.33,NaN,United Kingdom
15017,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,United Kingdom
16232,C537644,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:34:00,13474.79,NaN,United Kingdom
524601,C580604,AMAZONFEE,AMAZON FEE,-1,2011-12-05 11:35:00,11586.50,NaN,United Kingdom
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom


The highest UnitPrice are caused by special transaction fees such as amazon fees and Manual, rather than regular product sales.

--> For product sales analysis, non-product transactions such as Amazon fees, manual adjustments and bad debt adjustments should be excluded. Amazon Fee, Postage, DOTCOM POSTAGE, Manual, Adjust bad debt

### Q9. How many transactions belong to these non-product categories?

In [59]:
df[df["Description"].str.contains(
    "AMAZON FEE|POSTAGE|Manual|Adjust bad debt|DOTCOM POSTAGE",
    case=False, # case sensitivity
    na=False # NaN values are ignores
    )]["Description"].value_counts()

Description
POSTAGE            1252
DOTCOM POSTAGE      709
Manual              572
AMAZON FEE           34
Adjust bad debt       3
Name: count, dtype: int64

In [60]:
non_products = df[df["Description"].str.contains(
    "AMAZON FEE|POSTAGE|Manual|Adjust bad debt|DOTCOM POSTAGE",
    case=False, # case sensitivity
    na=False # NaN values are ignores
    )]

non_products = pd.DataFrame(non_products)

non_products["Revenue"] = non_products["Quantity"] * non_products["UnitPrice"]

non_products["Revenue"].sum()

np.float64(-28778.08000000001)

In [61]:
non_products.groupby(by="Description")["Revenue"].agg(["count", "sum"])

,count,sum
Description,,
AMAZON FEE,34,-221520.50
Adjust bad debt,3,-11062.06
DOTCOM POSTAGE,709,206245.48
Manual,572,-68671.64
POSTAGE,1252,66230.64


In [62]:
non_products_description = non_products["Description"].unique().tolist()

Several non-product transactions were identified, including Amazon fees, postage, manual adjustments and bad debt adjustments. These transactions can significantly affect revenue calculations and should therefore be handled separately from product sales.

### Q10. What are the original transactions of the cancelled ones?

In [63]:
df[df["InvoiceNo"].str.startswith("C")]

df.head()

cancel = df[df["InvoiceNo"] == "C536391"].iloc[0]

matches = df[
    (df["StockCode"] == cancel["StockCode"]) &
    (df["CustomerID"] == cancel["CustomerID"]) &
    (df["Quantity"] == -cancel["Quantity"]) &
    (df["UnitPrice"] == cancel["UnitPrice"]) &
    (df["InvoiceDate"] < cancel["InvoiceDate"])
]

print("Cancelled Transaktion: ")
display(cancel)

print("Original Transaction: ")
display(matches)



Cancelled Transaktion: 


InvoiceNo                             C536391
StockCode                               22556
Description    PLASTERS IN TIN CIRCUS PARADE 
Quantity                                  -12
InvoiceDate               2010-12-01 10:24:00
UnitPrice                                1.65
CustomerID                            17548.0
Country                        United Kingdom
Name: 235, dtype: object

Original Transaction: 


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


### Summary of Exploration

For the next step of cleaning the dataset, we have some rows, that we have to delete from our Dataset.

1. Duplicated rows: 5268
2. Canceled transactions: 9288
3. negative quantity: 10624 - 9288 = 1336
4. non-product transactions: AMAZON FEE, Manual
5. missing customerIDs: 135.080
6. missing descriptions: 1.454  

## Phase 2 – Data Cleaning

We need to sort out, which transaction went through and later got cancelled. These transactions are still in the dataset.

In [64]:
# first we gonna change the CustomerID from a float to an integer

df["CustomerID"] = pd.to_numeric(df["CustomerID"], errors="coerce") # Fehlerhafte Werte werden in NaN umgewandelt

df["CustomerID"] = df["CustomerID"].fillna(0).astype(int)

# remove duplicates

print(len(df))

df = df.drop_duplicates().copy()

print(f"Amount of duplicates: {df.duplicated().sum()}")
print(len(df))

## remove non-product transactions

df_sales = df[~df["Description"].str.contains(
    "AMAZON FEE|POSTAGE|Manual|Adjust bad debt|DOTCOM POSTAGE",
    case=False,
    na=False
)].copy() # dieses sucht eben nach allen non-product transactions und am ende negieren wir das, also nur product transactions

print(len(df_sales))

# remove canceled transactions

df_sales[df_sales["InvoiceNo"].str.startswith("C")]

541909
Amount of duplicates: 0
536641
534076


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
...,...,...,...,...,...,...,...,...
540448,C581490,22178,VICTORIAN GLASS HANGING T-LIGHT,-12,2011-12-09 09:57:00,1.95,14397,United Kingdom
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315,United Kingdom


In [ ]:
df_sales[(df_sales["CustomerID"] == 17548) & (df_sales["Description"].str.strip() == "PLASTERS IN TIN CIRCUS PARADE")] # mit str.strip(), werden alle Leerzeichen ausgeschnitten

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom


In [72]:
df_sales[(df_sales["CustomerID"] == 17548)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548,United Kingdom
165024,550755,22585,PACK OF 6 BIRDY GIFT TAGS,24,2011-04-20 12:01:00,1.25,17548,United Kingdom
165025,550755,22082,RIBBON REEL STRIPES DESIGN,10,2011-04-20 12:01:00,1.65,17548,United Kingdom
165026,550755,22081,RIBBON REEL FLORA + FAUNA,10,2011-04-20 12:01:00,1.65,17548,United Kingdom


In [83]:
df[df["InvoiceNo"].str.startswith("C")]

df.head()

cancel = df[df["InvoiceNo"] == "C536391"].iloc[0]

matches = df[
    (df["StockCode"] == cancel["StockCode"]) &
    (df["CustomerID"] == cancel["CustomerID"]) &
    (df["Quantity"] == -cancel["Quantity"]) &
    (df["UnitPrice"] == cancel["UnitPrice"]) &
    (df["InvoiceDate"] < cancel["InvoiceDate"])
]

print("Cancelled Transaktion: ")
display(cancel)

print("Original Transaction: ")
display(matches)

Cancelled Transaktion: 


InvoiceNo                             C536391
StockCode                               22556
Description    PLASTERS IN TIN CIRCUS PARADE 
Quantity                                  -12
InvoiceDate               2010-12-01 10:24:00
UnitPrice                                1.65
CustomerID                              17548
Country                        United Kingdom
Name: 235, dtype: object

Original Transaction: 


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


In [ ]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [86]:
print(df_sales.dtypes)

InvoiceNo                 str
StockCode                 str
Description               str
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID              int64
Country                   str
dtype: object
